sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -309,53 +306,112 @@ private IndexItemResponse indexItem(ShardUpsertRequest request,
+                                       @Nullable InsertSourceGen insertSourceGen) throws Exception {
+        assert insertSourceGen != null : "InsertSourceGen must not be null";
+        BytesReference rawSource;
+        Map<String, Object> source = null;
+        try {
+            // This optimizes for the case where the insert value is already string-based, so we can take directly
+            // the rawSource
+            if (insertSourceGen instanceof FromRawInsertSource) {
+                rawSource = insertSourceGen.generateSourceAndCheckConstraintsAsBytesReference(item.insertValues());
+            } else {
+                source = insertSourceGen.generateSourceAndCheckConstraints(item.insertValues());
+                rawSource = BytesReference.bytes(XContentFactory.jsonBuilder().map(source));
             }
-        } else {
-            Doc currentDoc = getDocument(indexShard, item.id(), item.version(), item.seqNo(), item.primaryTerm());
-
-            Map<String, Object> updatedSource = updateSourceGen.generateSource(
-                currentDoc,
-                item.updateAssignments(),
-                item.insertValues()
-            );
-
-            if (item.returnValues() != null) {
-                updatedDoc = currentDoc.withUpdatedSource(updatedSource);
+        } catch (IOException e) {
+            throw ExceptionsHelper.convertToElastic(e);
+        }
+        item.source(rawSource);
+
+        long version = request.duplicateKeyAction() == DuplicateKeyAction.OVERWRITE ? Versions.MATCH_ANY : Versions.MATCH_DELETED;
+        long seqNo = SequenceNumbers.UNASSIGNED_SEQ_NO;
+        long primaryTerm = SequenceNumbers.UNASSIGNED_PRIMARY_TERM;
+
+        Engine.IndexResult indexResult = index(item, indexShard, isRetry, seqNo, primaryTerm, version);
+        Object[] returnvalues = null;
+        if (returnGen != null) {
+            // This optimizes for the case where the insert value is already string-based, so only parse the source
+            // when return values are requested
+            if (source == null) {
+                source = JsonXContent.jsonXContent.createParser(
+                    NamedXContentRegistry.EMPTY,
+                    DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                    BytesReference.toBytes(rawSource)).map();
             }
+            returnvalues = returnGen.generateReturnValues(
+                // return -1 as docId, the docId can only be retrieved by fetching the inserted document again, which
+                // we want to avoid. The docId is anyway just valid with the lifetime of a searcher and can change afterwards.
+                new Doc(
+                    -1,
+                    indexShard.shardId().getIndexName(),
+                    item.id(),
+                    indexResult.getVersion(),
+                    indexResult.getSeqNo(),
+                    primaryTerm,
+                    source,
+                    () -> ""
```

Comment:
why not passing in the `rawSource` here as well? don't we want to support `_raw` as returning for performance reasons? If this was already discussed, sorry. If so a reasoning comment would be helpful ;)

Reply:
Yes, lets do it!